src/audio/aec/aec.c

```diff
@@ -0,0 +1,393 @@
+#include <sof/audio/pipeline.h>
+#include <sof/audio/aec/aec.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#define AEC_BLK_LENGTH 4
+#define AEC_MAX_SOURCES 2 /* Microphone and reference streams */
+
+static const struct comp_driver comp_aec;
+
+/* UUID 2ca424c0-7e1c-4b0a-afca-43de944705d3 */
+DECLARE_SOF_RT_UUID("aec", aec_uuid, 0x2ca424c0, 0x7e1c, 0x4b0a, 0xaf, 0xca,
+		    0x43, 0xde, 0x94, 0x47, 0x05, 0xd3);
+
+DECLARE_TR_CTX(aec_tr, SOF_UUID(aec_uuid), LOG_LEVEL_INFO);
+
+/* Generic processing */
+
+#if CONFIG_FORMAT_S16LE
+
+static void aec_s16_default(struct comp_dev *dev, const struct audio_stream **sources,
+			    struct audio_stream *sink, int frames, int ref_active)
+
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int16_t mix;
+	int16_t *x;
+	int16_t *y;
+	int16_t *r;
+	int in_idx;
+	int ref_idx;
+	int ref_nch;
+	int i;
+	int j;
+	int nch = sources[0]->channels;
+	int out_idx = 0;
+
+	if (ref_active)
+		ref_nch = sources[1]->channels;
+
+	/* Input from selected input and reference channels, same output to every channel */
+	in_idx = cd->source_channel;
+	ref_idx = cd->reference_channel;
+	for (i = 0; i < frames; i++) {
+		x = audio_stream_read_frag_s16(sources[0], in_idx);
+		mix = *x;
+		in_idx += nch;
+		if (ref_active) {
+			r = audio_stream_read_frag_s16(sources[1], ref_idx);
+			ref_idx += ref_nch;
+			mix = sat_int16(mix - *r);
+		}
+
+		for (j = 0; j < nch; j++) {
+			y = audio_stream_write_frag_s16(sink, out_idx);
+			*y = mix;
+			out_idx++;
+		}
+	}
+}
+
+#endif
+
+/* Processing functions table */
+const struct aec_func_map aec_fnmap[] = {
+#if CONFIG_FORMAT_S16LE
+	{ SOF_IPC_FRAME_S16_LE, aec_s16_default },
+#endif
+};
+
+const size_t aec_fncount = ARRAY_SIZE(aec_fnmap);
+
+/*
+ * End of AEC setup code. Next the standard component methods.
+ */
+
+static struct comp_dev *aec_new(const struct comp_driver *drv, struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct comp_data *cd;
+	struct sof_ipc_comp_process *aec;
+	struct sof_ipc_comp_process *ipc_aec = (struct sof_ipc_comp_process *)comp;
+	int ret;
+
+	comp_cl_info(&comp_aec, "aec_new()");
+
+	dev = comp_alloc(drv, COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	aec = COMP_GET_IPC(dev, sof_ipc_comp_process);
+	ret = memcpy_s(aec, sizeof(*aec), ipc_aec, sizeof(struct sof_ipc_comp_process));
+	assert(!ret);
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	cd->model_handler = comp_data_blob_handler_new(dev);
```

Comment:
I think I'll setup up clear text enum controls instead. AEC needs basically only enable/disable. Maybe coarse controls for few milliseconds of offset in reference delay, and maybe coarse control for room reverberation time. The sub-features like gain control enable/disable, residual echo conceal might use controls too.

Reply:
if its only on, we should probably use a switch rather than an enum